### PR TITLE
Overhaul stalled stream protection and add upload support

### PR DESCRIPTION
**Pull request by jdisanti**
This PR overhauls the existing stalled stream protection with a new algorithm, and also adds support for minimum throughput on upload streams. The new algorithm adds support for differentiating between the user or the server causing the stall, and not timing out if it's the user causing the stall. This will fix timeout issues when a customer makes remote service calls in between streaming pieces of information.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Comment by jdisanti** [comment]
Results of running the [stalled stream protection performance test](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/aws-smithy-runtime/tests/stalled_stream_performance.rs) in release mode.

```
Old impl without:  577.0 ± 2.9 ms
Old impl with:     584.1 ± 5.6 ms

New impl without:  577.1 ± 3.2 ms
New impl with:     583.3 ± 3.1 ms
```

These numbers are across six separate runs of the performance test each, averaged, and with a standard deviation applied. The performance is roughly the same, with maybe a miniscule amount of improvement.

The new implementation uses 80 bytes of memory for the log instead of 1 kilobyte though.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:102]: _nitpick_

Should we avoid calling this `Cell` because of the connotations that that has for Rust programmers? Typically `Cell` means some sort of pointer type.

As an alternative, I offer the equally meaningless `Bin`

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:159]: Is this our first usage of a GAT? If so, 🎊

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:177]:
```suggestion
    /// ## Panics
    ///
    /// The buffer MUST have at least one cell it it before this is called or it will panic.
```

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:265]: Great docs! These are really first class John.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:356]: lol I'm always making these explicit so that we don't import things in the parent scope that only get used in the child scope. We're covering each other's paint.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:342]: Consider adding a comment saying why this mean the report is incomplete.

> If the buffer is less than half full of logs, then the report is considered incomplete.
> This is because...

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:503]: alternatively,
```suggestion
        for i in (50..=850).step_by(100) {
                logs.push_pending(start + Duration::from_millis(i));
        }
```

Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:503]: I didn't test this code so don't hit that commit button

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (Velfi) [inline rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:57]: 😆

**Review by Velfi — APPROVED** [review]

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:356]: I don't want to think about imports when writing tests 😆

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:503]: 350 is skipped, so it's not quite equivalent. Perhaps it deserves a comment.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:342]: Now that I think about it, I'm not really sure why I chose to compare against half. If any of the cells are empty, it should be incomplete. Revising that...

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:159]: assuming you mean "const generic", I think yes!

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:57]: Need to make sure size_hint is passed through.

File: aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs
Comment (rcoh) [inline aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs:98]: by default is it not enabled for upload? why is this required?

File: aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs
Comment (rcoh) [inline aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs:57]: recommend using http_body 1.0 for bodies going forward

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt
Comment (rcoh) [inline codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt:127]: do you think we need a BMV to enable it on request bodies?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:155]: could consider using http_body 1.0 here as well

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (rcoh) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:57]: yeah, we may want an integration test that verifies that size hint and end-stream hint go all the way up and down the stack so we don't have any bodies that forget to delegate them.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:20]: may want to consider returning an enum instead of a bool

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:101]: So, I would have thought sleep_fut wasn't required anymore since we are using the ThroughputCheckFuture externally. But if I remove it, download_stalls and download_too_slow fail. Do you understand why? 

Maybe it's just a matter of settings? I wonder if we could refactor that so we can remove that complexity from here.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:174]: seems like we could slightly refactor the API to be something like:

`this.throughput.record_poll(poll_result)`

```
fn record_poll(&mut self, p: Poll<Bytes>) -> Poll<Bytes>
```

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs:138]: deprecation should probably refer to `check_window` since that appears to be the replacement?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:116]: So a bin transitions from `TransferredBytes` into `Pending` if the last thing that happened in the bin was `Poll::Pending`?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:248]: this is the number of throughput windows?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:291]: need an example of some sort to understand what resolution means in this context. Is it the width of a bin?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:165]: what if you add a third bin option `Untouched` or something? Then we can remove this length logic which makes me a little uncomfortable due to off-by-one risk

**Review by rcoh — COMMENTED** [review]
I got through about half of it, will dig in more later

**Review by ysaito1001 — APPROVED** [review]
This is one of those wow PRs, excellent work!

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (ysaito1001) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:291]: I think it is. A resolution is the amount of time each cell is responsible for.

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:97]: I don't think we can yet?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:101]: This is the download body, which _does not_ use ThroughputCheckFuture since it's not the orchestrator that's waiting on it. Rather, the customer is waiting on it in their code directly since it is the response body.

File: rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs:28]: it would be nice if you could "loud" this from the API instead of needing to use an environment variable.

File: rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs:45]: should we just pass the contents to `with_env_filter` instead? Then you don't need to parse match etc.

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:60]: I assume `NEAT` here just really means "test data"?

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:98]: we probably need a module of test bodies. we have a bunch of private ones everywhere

File: CHANGELOG.next.toml
Comment (rcoh) [inline CHANGELOG.next.toml:27]: this is one super long sentence

File: aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs
Comment (jdisanti) [reply aws/sdk/integration-tests/s3/tests/stalled-stream-protection.rs:98]: Just being explicit.

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt:127]: Yeah, I'm not sure. I guess it depends on whether or not we consider this a bug fix?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs:138]: I'm not sure I want to make check_window actually configurable or not yet. I was leaning towards not exposing it until a use-case came up for it.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:116]: Yeah, `Pending` is the highest priority in the `Bin::merge` function below.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:127]: In your hypothetical case, it depends on what `other` is. `this` doesn't have a lot of bearing on what is chosen until the last few match arms. But the gist of it is what the comment above the match says: Pending > TransferredBytes > NoPolling > Empty.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:165]: I had that approach originally, but it only moves the problem. Going that way, you have to track the time that the tracking started so that you know which cells can be flipped to NoPolling.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:248]: Should rename this to `BIN_COUNT`.

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:60]: Pretty much 😄

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/options.rs:138]: Offline discussion: Improve deprecation messaging to indicate that this configuration is not currently available and to file an issue if it's needed.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:126]:
```
(other, Pending(this)) => Pending(other.bytes() + this),
```

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs:57]: Fixed and integration tested in 73635049dc5bc5fabb603292fec92d5d803e9706

File: codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt
Comment (jdisanti) [reply codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/config/StalledStreamProtectionConfigCustomization.kt:127]: Filed https://github.com/smithy-lang/smithy-rs/issues/3510 to track the BMV change.

File: rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs:45]: It's not a bad idea, but as I was attempting it, I realized we're kind of overloading the purpose of this test utility. We should probably add a separate `show_test_logs()` function that does the env filtering.

File: rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs
Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime/tests/stalled_stream_common.rs:96]: Punting on this in this PR.

File: CHANGELOG.next.toml
Comment (jdisanti) [reply CHANGELOG.next.toml]: Split into two changelog entries.

File: CHANGELOG.next.toml
Comment (rcoh) [inline CHANGELOG.next.toml:27]:
```suggestion
message = "Stalled stream protection on downloads will now only trigger if the upstream source was too slow. Previously, stalled stream protection could be erroneously triggered if the customer was slowly consuming the stream slower than the minimum speed limit."
```

File: aws/sdk/integration-tests/s3/tests/body_size_hint.rs
Comment (rcoh) [inline aws/sdk/integration-tests/s3/tests/body_size_hint.rs:65]: you can avoid the need to make a entire new struct by using `infallible_client_fn`

File: aws/sdk/integration-tests/s3/tests/body_size_hint.rs
Comment (rcoh) [inline aws/sdk/integration-tests/s3/tests/body_size_hint.rs:106]: I think this can use `capture_request` (no need for adhoc struct)

File: rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/test_util/capture_test_logs.rs:31]: I think we should add a way to loudly show test logs without having to set an env var for convenience (e.g. when running tests w/ intellij)

Could make a `loud` argument to the variable, or add a boolean arg to the function?

Or keep `show_test_logs` as the loud one?

File: CHANGELOG.next.toml
Comment (rcoh) [inline CHANGELOG.next.toml:32]: the changelog entries should also discuss _how_ to enable protection for uploads before it becomes a new BMV

**Review by rcoh — APPROVED** [review]

**Pull request merged**
